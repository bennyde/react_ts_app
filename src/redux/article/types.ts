
// Action types
export const REQUEST_ARTICLE: any = 'REQUEST_ARTICLE'
export const RECEIVE_ARTICLE: any = 'RECEIVE_ARTICLE'
export const RECEIVE_ARTICLE_ERROR: any = 'RECEIVE_ARTICLE_ERROR'

// Action interfaces
interface RequestArticleAction {
    type: typeof REQUEST_ARTICLE
    payload: string
}

interface ReceiveArticleAction {
    type: typeof RECEIVE_ARTICLE
    payload: Article
}

interface ReceiveArticleErrorAction {
    type: typeof RECEIVE_ARTICLE_ERROR
    payload: Error
}

export type ArticleActionTypes = ReceiveArticleAction | RequestArticleAction | ReceiveArticleErrorAction

// State interfaces
export interface Article {
    id: string
    title: string
    excerpt: string
    content: string
    author: string
    image: string
    date: string
    category: string
    viewed: Number
}

export interface Error {
    message: string
}

export interface ArticleState {
    id: string
    article: Article
    isFetching: boolean
    error: Error
}

