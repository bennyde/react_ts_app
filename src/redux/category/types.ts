import { Article } from '../article/types'

// Action types
export const REQUEST_CATEGORY: string = 'REQUEST_CATEGORY'
export const RECEIVE_CATEGORY: string = 'RECEIVE_CATEGORY'
export const RECEIVE_CATEGORY_ERROR: string = 'RECEIVE_CATEGORY_ERROR'

interface RequestCategoryAction {
    type: typeof REQUEST_CATEGORY
    payload: string
}

interface ReceiveCategoryAction {
    type: typeof RECEIVE_CATEGORY
    payload: Array<Article>
}

interface ReceiveCategoryErrorAction {
    type: typeof RECEIVE_CATEGORY_ERROR
    payload: Error
}

export type CategoryActionTypes = ReceiveCategoryAction | RequestCategoryAction | ReceiveCategoryErrorAction

// State
export interface Category {
    name: string
    articles: Array<Article>
}

export interface Error {
    message: string
}

export interface CategoryState {
    category: Category
    isFetching: boolean
    error: Error
}

