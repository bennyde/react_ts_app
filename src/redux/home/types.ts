import { Article } from '../article/types'

// Action types
export const REQUEST_ARTICLES: any = 'REQUEST_ARTICLES'
export const RECEIVE_ARTICLES: any = 'RECEIVE_ARTICLES'
export const RECEIVE_ARTICLES_ERROR: any = 'RECEIVE_ARTICLES_ERROR'

interface RequestArticlesAction {
    type: typeof REQUEST_ARTICLES,
    payload: any
}

interface ReceiveArticlesAction {
    type: typeof RECEIVE_ARTICLES
    payload: Article[]
}

interface ReceiveArticlesErrorAction {
    type: typeof RECEIVE_ARTICLES_ERROR
    payload: Error
}

export type ArticlesActionTypes = ReceiveArticlesAction | RequestArticlesAction | ReceiveArticlesErrorAction


export interface Error {
    message: string
}

export interface ArticlesState {
    articles: Article[]
    isFetching: boolean
    error: Error,
    page: number,
    posts_per_page: number
}

