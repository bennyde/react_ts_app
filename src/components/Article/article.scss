.article {
    max-width: 82rem;
    margin: 0 auto;
    img {
        width: 100%;
        max-width: 80rem;
    }
}
.article-content h1::first-letter {
    text-transform: capitalize;
}