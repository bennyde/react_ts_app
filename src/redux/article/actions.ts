import { Article, Error, REQUEST_ARTICLE, RECEIVE_ARTICLE, RECEIVE_ARTICLE_ERROR } from "./types"

export function requestArticle(id: string) {
    console.log('ACTION REQUEST_ARTICLE CALLED id ' + id)
    return {
        type: REQUEST_ARTICLE,
        payload: id
    }
}

export function receiveArticle(article: Article) {
    console.log('ACTION RECEIVE_ARTICLE CALLED title: ' + article.title)
    return {
        type: RECEIVE_ARTICLE,
        payload: article
    }
}

export function receiveArticleError(error: Error) {
    //console.log('ACTION RECEIVE_ARTICLE_ERROR CALLED error: ' + JSON.stringify(error))
    return {
        type: RECEIVE_ARTICLE_ERROR,
        payload: error
    }
}
