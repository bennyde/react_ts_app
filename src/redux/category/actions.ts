import { Category, Error, REQUEST_CATEGORY, RECEIVE_CATEGORY, RECEIVE_CATEGORY_ERROR } from "./types"

export function requestCategory(name: string) {
    //console.log('ACTION REQUEST_CATEGORY CALLED category ' + name)
    return {
        type: REQUEST_CATEGORY,
        payload: name
    }
}

export function receiveCategory(category: Category) {
    //console.log('ACTION RECEIVE_ARTICLE CALLED title: ' + article.title)
    return {
        type: RECEIVE_CATEGORY,
        payload: category
    }
}

export function receiveCategoryError(error: Error) {
    //console.log('ACTION RECEIVE_ARTICLE_ERROR CALLED error: ' + JSON.stringify(error))
    return {
        type: RECEIVE_CATEGORY_ERROR,
        payload: error
    }
}
