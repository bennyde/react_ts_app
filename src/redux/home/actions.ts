import { Article } from '../article/types'
import { Error, REQUEST_ARTICLES, RECEIVE_ARTICLES, RECEIVE_ARTICLES_ERROR } from "./types"

export function requestArticles(page: number = 1, posts_per_page: number = 12) {
    //console.log('ACTION REQUEST_ARTICLES CALLED')
    return {
        type: REQUEST_ARTICLES,
        payload: { page, posts_per_page }
    }
}

export function receiveArticles(articles: Article[]) {
    //console.log('ACTION RECEIVE_ARTICLE CALLED title: ' + articles)
    return {
        type: RECEIVE_ARTICLES,
        payload: articles
    }
}

export function receiveArticlesError(error: Error) {
    //console.log('ACTION RECEIVE_ARTICLE_ERROR CALLED error: ' + JSON.stringify(error))
    return {
        type: RECEIVE_ARTICLES_ERROR,
        payload: error
    }
}
