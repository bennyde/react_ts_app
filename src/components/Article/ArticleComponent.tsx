import * as React from "react"
import { useSelector, useDispatch } from 'react-redux'
import { ArticleState, Error } from '../../redux/article/types'
import { requestArticle } from '../../redux/article/actions'
import { RouteComponentProps } from 'react-router-dom';

const ArticleComponent = (props: RouteComponentProps) => {
	let title, excerpt, content, author, date, image

	const articleState: ArticleState = useSelector((state: any) => state.articleState)
	if (articleState && articleState.article) {
		({ title, excerpt, content, author, date, image } = articleState.article)
	}

	const dispatch = useDispatch()

	React.useEffect(() => {
		try {
			const params: any = props.match.params
			dispatch(requestArticle(params.id))
		} catch (e) {
			console.log('Error: ' + e)
		}
	}, [])

	if (articleState && articleState.error) {
		const error: Error = articleState.error
		return (
			<div className="section article theme-santafe">
				<p>ERROR</p>
				<p>{error.message}</p>
			</div>
		)
	}

	if (articleState && articleState.isFetching) {
		return (
			<div className="section article theme-santafe">
				<p>LOADING...</p>
			</div>
		)
	}

	return (
		<div className="section article">
			<div className="article-content">
				<h1>{title}</h1>
				<img src={image} alt="image" />
				<p>{author}</p>
				<p>{date}</p>
				<p>{excerpt}</p>
				<p>{content}</p>
			</div>
		</div>
	)
}

export default ArticleComponent
