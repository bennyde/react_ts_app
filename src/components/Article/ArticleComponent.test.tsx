// xit - skip test
// fit - focus on only this test

// .skip - skip test
// .only - only this test

import * as React from "react"
// import * as ReactTestUtils from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import * as renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import ArticleComponent from "./ArticleComponent"
import { Article, Error } from "../../redux/article/types"
// import { fetchArticle } from "../../store/article/actions"
import { mount, shallow, ShallowWrapper } from "enzyme"
import { RouteComponentProps } from 'react-router-dom'
// import store from "../../redux"
import { requestArticle, receiveArticle, receiveArticleError } from '../../redux/article/actions'
import toJSON from "enzyme-to-json"
import { articles } from '../../sagas/fake_data'

//const act = { renderer };

const props: any = {
    "history": {
        "length": 2,
        "action": "PUSH",
        "location": {
            "pathname": "/article/1",
            "search": "",
            "hash": ""
        }
    },
    "location": {
        "pathname": "/article/1",
        "search": "",
        "hash": ""
    },
    "match": {
        "path": "/article/:id",
        "url": "/article/1",
        "isExact": true,
        "params": {
            "id": "1"
        }
    }
}

const props_2: any = {
    "history": {
        "length": 2,
        "action": "PUSH",
        "location": {
            "pathname": "/article/245",
            "search": "",
            "hash": ""
        }
    },
    "location": {
        "pathname": "/article/245",
        "search": "",
        "hash": ""
    },
    "match": {
        "path": "/article/:id",
        "url": "/article/245",
        "isExact": true,
        "params": {
            "id": "245"
        }
    }
}

test('Test Article object', () => {
    expect(1).toEqual(1)
})

describe.skip('<ArticleComponent />', () => {
    it('renders ArticleComponent with provided props', () => {
        shallow(<Provider store={store}><ArticleComponent {...props} /></Provider >)
    })
    it('matches the snapshot', () => {
        const wrapper: ShallowWrapper = shallow(<Provider store={store}><ArticleComponent {...props} /></Provider >)
        expect(toJSON(wrapper, { noKey: false, mode: 'deep' })).toMatchSnapshot()
        //expect(wrapper.find('ArticleComponent').dive()
    })
    it('updates via props', () => {
        const wrapper: ShallowWrapper = shallow(<Provider store={store}><ArticleComponent {...props} /></Provider >)
        expect(toJSON(wrapper)).toMatchSnapshot()
        wrapper.setProps({ children: <ArticleComponent {...props_2} /> });
        expect(toJSON(wrapper)).toMatchSnapshot()
    })
    it('mounts the component', () => {
        const wrapper = mount(<Provider store={store}><ArticleComponent {...props} /></Provider >)
        console.log(wrapper.debug())
        expect(wrapper.find('ArticleComponent').exists()).toBe(true)
        expect(wrapper.find('ArticleComponent')).toHaveLength(1)
        // expect(wrapper.find('#main').exists()).toBe(true)
    })
})


let store: any
describe.skip('<ArticleComponent /> test actions', () => {

    let component: any
    const mockStore = configureStore([]);
    beforeEach(() => {
        store = mockStore({
            articleState: {
                id: '1',
                article: articles[0],
                isFetching: false,
                error: null
            }
        })
        renderer.act(() => {
            component = renderer.create(
                <Provider store={store}>
                    <ArticleComponent {...props} />
                </Provider>)
        })
    })
    test('should render with given state from Redux store', () => {
        const state = store.getState()
        console.log('state ' + JSON.stringify(state))
        expect(component.toJSON()).toMatchSnapshot()
    })
    test('should dispatch REQUEST_ARTICLE action', () => {
        store.dispatch(requestArticle('1'))
        const actions = store.getActions()
        const expectedPayload = { type: 'REQUEST_ARTICLE', payload: '1' }
        expect(actions[0]).toEqual(expectedPayload)
    })
    test('should dispatch RECEIVE_ARTICLE action', () => {
        store.dispatch(receiveArticle(articles[0]))
        const actions = store.getActions()
        const expectedPayload = { type: 'RECEIVE_ARTICLE', payload: articles[0] }
        expect(actions[1]).toEqual(expectedPayload)
    })
    test('should dispatch RECEIVE_ARTICLE_ERROR action', () => {
        const error: Error = { message: 'Error occurred' }
        store.dispatch(receiveArticleError(error))
        const actions = store.getActions()
        const expectedPayload = { type: 'RECEIVE_ARTICLE_ERROR', payload: error }
        expect(actions[1]).toEqual(expectedPayload)
    })
    it('should update with given state from Redux store', () => {
        store.dispatch(receiveArticle(articles[0]))
        const state = store.getState()
        console.log('state ' + JSON.stringify(state))
        expect(toJSON(component)).toMatchSnapshot()
    })
})


// let component: any
// describe('<ArticleComponent /> connected React-Redux Component', () => {
//     const mockStore = configureStore([]);
//     beforeEach(() => {
//         store = mockStore({
//             articleState: {
//                 id: '1',
//                 article: articles[0],
//                 isFetching: false,
//                 error: null
//             }
//         })
//         component = mount(
//             <Provider store={store}>
//                 <ArticleComponent {...props} />
//             </Provider>
//         )
//     })
//     it('should render with given state from Redux store', () => {
//         expect(toJSON(component)).toMatchSnapshot()
//     })
//     it('should update with given state from Redux store', () => {
//         store.dispatch(receiveArticle(articles[0]))
//         const state = store.getState()
//         console.log('state ' + JSON.stringify(state))
//         expect(toJSON(component)).toMatchSnapshot()
//     })
// })