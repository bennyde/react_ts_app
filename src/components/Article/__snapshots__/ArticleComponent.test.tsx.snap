// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleComponent /> test actions should render with given state from Redux store 1`] = `
<div
  className="section article theme-santafe"
>
  <div
    className="article-content"
  >
    <h1>
      Top skiing weekends
    </h1>
    <img
      alt="image"
      src="http://localhost:8080/public/assets/chamonix.jpg"
    />
    <p>
      John Smith
    </p>
    <p>
       Thursday, August 8th, 2019
    </p>
    <p>
      Whether you’ve business nearby or have simply had a trying week at work, what better way to relax than to visit one of the world’s greatest ski resorts?
    </p>
    <p>
      Chamonix, France - Located a mere 88km from Geneva airport, the infamous resort of Chamonix is renowned the world over for its adrenaline inducing slopes and postcard picture scenery. As if the daylight hour activities weren’t enough, Chamonix also boasts a lively nightlife and host of fine dining opportunities. 
St Anton, Austria - Widely cited as the best ski resort in Austria, St Anton is largely intended for the experienced skier and offers an unparalleled number and variety of slopes. However, if you’re feeling less adventurous then the many slope-side bars and restaurants offer an opportunity to kick back and observe the vibrant ski scene. 
Mayrhofen, Austria - A short 65km from Innsbruck Airport lies Mayrhofen, a 630m high village with a buzzy atmosphere and a great many skiing opportunities for intermediate and advanced skiers alike. The destination is especially accommodating to snowboarders, who customarily gather in April for the annual Snowbombing festival.
    </p>
  </div>
</div>
`;

exports[`<ArticleComponent /> test actions should update with given state from Redux store 1`] = `undefined`;
